Environment variable registration, deregistration (addresses #3227)

dotsdl opened the pull request:
<!--- Thanks for opening a PR on xonsh! Please include a news entry with your PR
to help keep our changelog up to date! There are instructions available here:
https://xon.sh/devguide.html#changelog -->

<!--- If there is specific issue / feature request that this PR is addressing,
please link to the corresponding issue by using the `#issuenumber` syntax.
Thanks again! -->

This PR addresses #3227.

Added `xonsh.environ.Env.register` and `xonsh.environ.Env.deregister` methods for registration, deregistration of new environment variables with types, default values, and docstrings. Currently registration/deregistration is not exposed to the user, so the addition of these gives the ability to add new variables with behavior like e.g. `PATH`.

dotsdl commented:
Needs at least:
1. explicit tests
2. recommendations for the way we expose the variable machinery in these functions.

scopatz commented on xonsh/environ.py:
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
+            Function to validate type.
+        ens_convert : func, optional
+            Function to convert variable from a string representation to its type.
+        ens_detype : func, optional
```
These could probably drop the `ens_` prefix

dotsdl replied:
Done!

scopatz commented on xonsh/environ.py:
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
+        defaultval : optional
+            Default value for variable. ``ValueError`` raised if type does not match
+            that specified by `vartype` (or `ens_validate`).
+        vardoc : str, optional
```
this could be called doc

dotsdl replied:
Done!

scopatz commented on xonsh/environ.py:
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
+        varname : str
+            Environment variable name to register. Typically all caps.
+        vartype : str, optional,  {'bool', 'str', 'path', 'int', 'float'}
+            Variable type. If not one of the available presets, use `ens_validate`,
```
These could probably drop `var`

dotsdl replied:
Done!

scopatz commented on xonsh/environ.py:
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
+        vartype : str, optional,  {'bool', 'str', 'path', 'int', 'float'}
+            Variable type. If not one of the available presets, use `ens_validate`,
+            `ens_convert`, and `ens_detype` to specify type behavior.
+        defaultval : optional
```
This could probably just be `default`

dotsdl replied:
Done!

scopatz commented on xonsh/environ.py:
```diff
@@ -1647,6 +1655,88 @@ def _repr_pretty_(self, p, cycle):
+
+        # set doc for envvar
+        if vardoc is not None:
+            self._docs[varname] = VarDocs(
```
Ideally I think we'd combine VarDocs and Ensurer into a single namedtuple

dotsdl replied:
Can do. Do we want to merge `DEFAULT_ENSURERS` and `DEFAULT_DOCS` along with this change (seems inevitable). Would we also want to merge in `DEFAULT_VALUES` to this new namedtuple?

dotsdl replied:
ping @scopatz; do we want to merge together all the `DEFAULT_*` lazy objects?

scopatz replied:
Yes, I think we do.  I am so sorry for taking so long to get back to this here

dotsdl replied:
No worries, all good! I will proceed with implementation. I believe I can carve out some time next week.

scopatz replied:
Awesome! thanks @dotsdl!

scopatz commented:
Thanks for putting this in @dotsdl! Sorry I have had a couple of crazy weeks and didn't take a look sooner

gforsyth commented:
Hey @dotsdl !  Sorry for the radio silence -- what's the status on this?  I'm seeing `black` failures that should be pretty quick to fix, is anything else needed here?

dotsdl commented:
@gforsyth there is one unresolved question for @scopatz up above. I am happy to implement once I have clarity on what the desired end state is.

astronouth7303 commented:
@dotsdl @scopatz ping

dotsdl commented:
I believe I've got everything needed in place. Now need to address tests.

dotsdl commented:
I'm getting test failures that I seriously don't understand. If anyone has time to glance, I'll take any ideas you have for what might be happening here.

scopatz commented:
I think that the issue here is that the history replayer didn't get updated to use the new API. We could consider getting rid of it, since I am not sure anyone uses it.

gforsyth commented:
Hey @dotsdl -- let's just remove history replay.

dotsdl commented:
I believe this is ready to go. I'm not sure what's going on with the CI, but all tests pass for me locally here. This one has been on my list to revisit for a long time, so thank you for everyone's patience.

I've removed `replay.py` and `test_replay.py`, so no more history replay. I've also added some explicit tests for registration/deregistration of env variables. Please check that these tests align with your expectations for how this should work.

Anything additional needed for merge?

scopatz commented:
@xonsh/xore - this is ready for review! I approve of @dotsdl's changes, but I did some cleanup here, so if someone else wants to look at it, that would be good too

dotsdl commented:
Awesome, thank you @scopatz! I saw the flurry of activity and got excited. Thanks for helping me get it to the finish line!

scopatz commented:
Of course!

gforsyth commented:
:rocket: thanks @dotsdl and @scopatz !!

laloch commented:
Huh, really? I completely missed this one. This removes ensurers!

scopatz commented:
It changes the API so that so can specify EnvVar behavior in one place.

laloch commented:
Yes, I understand its purpose, but it also removes the possibility to restrict a variable to a programmable set of values.

scopatz commented:
It shouldn't.  The ensurers are still there.

laloch commented:
Oh, now I see! You're right. Sorry for the noise.

scopatz commented:
What do you think is missing @laloch?

laloch commented:
Everything is in place. I'm an idiot. Sorry. :facepalm:

scopatz commented:
:+1: Yay! I was scared for a moment!
